dask_geomodeling/raster/sources.py

```diff
@@ -14,7 +14,117 @@
 __all__ = ["MemorySource", "RasterFileSource"]
 
 
-class MemorySource(RasterBlock):
+class RasterSourceBase(RasterBlock):
+
+    @staticmethod
+    def process(process_kwargs):
+        mode = process_kwargs["mode"]
+
+        # handle empty requests
+        if mode == "empty_vals":
+            return
+        elif mode == "empty_time":
+            return {"time": []}
+        elif mode == "empty_meta":
+            return {"meta": []}
+
+        bands = process_kwargs["bands"]
+        length = bands[1] - bands[0]
+
+        # handle time requests
+        if mode == "time":
+            start = process_kwargs["start"]
+            delta = process_kwargs["delta"]
+            return {"time": [start + i * delta for i in range(length)]}
+
+        # open the dataset
+        source = process_kwargs.get("dataset")
+        if source is None:
+            url = process_kwargs["url"]
+            path = utils.safe_abspath(url)
+            source = gdal.Open(path)
+
+        # handle meta requests
+        if mode == "meta":
+            return {
+                "meta": [
+                    source.GetRasterBand(i + 1).GetMetadataItem("metadata")
+                    for i in range(bands[0], bands[1])
+                ]
+            }
+
+        # handle 'vals' requests
+        dtype = process_kwargs["dtype"]
+        bbox = process_kwargs["bbox"]
+        width = process_kwargs["width"]
+        height = process_kwargs["height"]
+        target_projection = process_kwargs["projection"]
+        target_no_data_value = process_kwargs["fillvalue"].item()
+
+        # return an empty array if 0-sized data was requested
+        if width == 0 or height == 0:
+            return np.empty((length, height, width), dtype=dtype)
+
+        # handle point request
+        if bbox[0] == bbox[2] or bbox[1] == bbox[3]:
+            source_geo_transform = utils.GeoTransform(source.GetGeoTransform())
+            ct = osr.CoordinateTransformation(
+                utils.get_sr(target_projection),
+                utils.get_sr(source.GetProjection()),
+            )
+            source_x, source_y, _ = ct.TransformPoint(bbox[0], bbox[1], 0.0)
+            source_i, source_j = source_geo_transform.get_indices(
+                ((source_x, source_y))
+            )
+            # create the result
+            shape = source.RasterCount, 1, 1
+            result = np.full(shape, target_no_data_value, dtype=dtype)
+            if (
+                0 <= source_i < source.RasterYSize
+                and 0 <= source_j < source.RasterXSize
+            ):
+                # read data into the result
+                source.ReadAsArray(int(source_j), int(source_i), 1, 1, result)
+
+            result = result[bands[0] : bands[1]]
+            # fill nan values if they popped up
+            result[~np.isfinite(result)] = target_no_data_value
```

Comment:
Not sure if this works for integers?

Reply:
It does, but I removed this statement nevertheless because it was a left-over from the earlier implementation that used `scipy.ndimage.zoom`.